<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from urdf/device.urdf.xacro         | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="twin_hammer">
  <baselink name="fc"/>
  <thrust_link name="thrust"/>
  <!-- <xacro:property name="pi" value="3.1415926535897931" /> -->
  <!-- dynamics -->
  <m_f_rate value="0"/>
  <!-- [N] -->
  <!-- <xacro:macro name="rotor" params="number"> -->
  <!-- dummy root link for KDL -->
  <link name="root">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.00001"/>
      <inertia ixx="0.000001" ixy="0.0" ixz="0.0" iyy="0.000001" iyz="0.0" izz="0.000002"/>
    </inertial>
  </link>
  <joint name="root_joint" type="fixed">
    <parent link="root"/>
    <child link="main_body"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <axis xyz="0 0 1"/>
  </joint>
  <!-- when fix main body  -->
  <link name="dummy_weight">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="100"/>
      <inertia ixx="100" ixy="100" ixz="100" iyy="100" iyz="100" izz="100"/>
    </inertial>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.5 0.5 1.0"/>
      </geometry>
    </collision>
  </link>
  <!-- <joint name="dummy_weight_joint" type="fixed">
    <parent link="dummy_weight"/>
    <child link="root"/>
    <origin rpy="0 0 0" xyz="0 0 1.0"/>
  </joint> -->
  <!-- ft_sensor_gazebo joint -->
  <joint name="root_to_dummyweight_joint" type="fixed">
    <parent link="dummy_weight"/>
    <child link="root"/>
    <origin rpy="0 0 0" xyz="0 0 1.0"/>
  </joint>
  <!-- main_body -->
  <link name="main_body">
    <inertial>
      <mass value="0.25"/>
      <origin rpy="0 0 0" xyz="0.1356 -0.0001 -0.0139"/>
      <inertia ixx="0.0002" ixy="0" ixz="-0.0007" iyy="0.0175" iyz="0" izz="0.0174"/>
    </inertial>
    <visual>
      <!-- <origin rpy="${pi/2} 0 ${pi/2}" xyz="-0.8 0 0"/> -->
      <origin rpy="1.5707963267948965 0 1.5707963267948965" xyz="-0.8 0 0"/>
      <geometry>
        <mesh filename="package://twin_hammer/urdf/mesh/main_body.dae" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>
    <collision>
      <!--
     <origin rpy="0 ${pi/2} 0" xyz="0 0 0"/>
     <geometry>
       <cylinder length="${rod_length}" radius="0.125" />
     </geometry>
     -->
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="1.0 0.15 0.15"/>
      </geometry>
    </collision>
  </link>
  <!-- spinal -->
  <link name="fc">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.00001"/>
      <inertia ixx="0.000001" ixy="0.0" ixz="0.0" iyy="0.000001" iyz="0.0" izz="0.000002"/>
    </inertial>
  </link>
  <joint name="fc_joint" type="fixed">
    <child link="fc"/>
    <parent link="main_body"/>
    <origin rpy="0 0 0" xyz="0.2341 0 0.0292"/>
    <axis xyz="0 0 1"/>
  </joint>
  <!-- id:1 is spinal side, id:2 is another side-->
  <!-- gimbal roll -->
  <joint name="gimbal1_roll" type="revolute">
    <limit effort="3" lower="-3.14" upper="3.14" velocity="12.5"/>
    <parent link="main_body"/>
    <child link="gimbal1_roll_module"/>
    <origin rpy="0 0 0" xyz="0.4638 0 0"/>
    <axis xyz="1 0 0"/>
    <dynamics damping="0.1" friction="0.0"/>
  </joint>
  <link name="gimbal1_roll_module">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.0151 0.0012 0.0233"/>
      <mass value="0.3335"/>
      <inertia ixx="0.0008" ixy="0" ixz="-0.0002" iyy="0.0009" iyz="0" izz="0.0003"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://twin_hammer/urdf/mesh/gimbal_roll_module.dae" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>
    <!-- box collision not work good for FCL
     <collision>
       <origin rpy="0 0 0" xyz="${gimbal_roll_x_offset}  0 0"/>
       <geometry>
         <box size="0.103 0.064 0.156"/>
       </geometry>
     </collision>
     -->
    <!-- cylinder collision geometry works well for FCL -->
    <collision>
      <!-- <origin rpy="${pi/2} 0 0" xyz="0 0 0"/> -->
      <origin rpy="1.5707963267948965 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.3" radius="0.10"/>
      </geometry>
    </collision>
  </link>
  <!-- gimbal pitch -->
  <joint name="gimbal1_pitch" type="revolute">
    <limit effort="3" lower="-3.14" upper="3.14" velocity="12.5"/>
    <parent link="gimbal1_roll_module"/>
    <child link="gimbal1_pitch_module"/>
    <origin rpy="0 0 0" xyz="0.0065 0 0.0265"/>
    <axis xyz="0 1 0"/>
    <dynamics damping="0.1" friction="0.0"/>
  </joint>
  <link name="gimbal1_pitch_module">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0.0005 -0.0092"/>
      <mass value="0.5276"/>
      <inertia ixx="0.0063" ixy="0.0" ixz="0.0" iyy="0.0004" iyz="0.0" izz="0.0062"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://twin_hammer/urdf/mesh/gimbal_pitch_module.dae" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>
  </link>
  <!-- rotor -->
  <joint name="rotor1" type="continuous">
    <limit effort="100.0" lower="2" upper="30" velocity="0.5"/>
    <parent link="gimbal1_pitch_module"/>
    <child link="thrust1"/>
    <origin rpy="0 0 0" xyz="0 0.1115 0"/>
    <axis xyz="0 0 1"/>
  </joint>
  <link name="thrust1">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.0001"/>
      <inertia ixx="0.00001" ixy="0.0" ixz="0.0" iyy="0.00001" iyz="0.0" izz="0.00002"/>
    </inertial>
  </link>
  <joint name="rotor3" type="continuous">
    <limit effort="100.0" lower="2" upper="30" velocity="0.5"/>
    <parent link="gimbal1_pitch_module"/>
    <child link="thrust3"/>
    <origin rpy="0 0 0" xyz="0 -0.1115 0"/>
    <axis xyz="0 0 1"/>
  </joint>
  <link name="thrust3">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.0001"/>
      <inertia ixx="0.00001" ixy="0.0" ixz="0.0" iyy="0.00001" iyz="0.0" izz="0.00002"/>
    </inertial>
  </link>
  <!-- special part of duct fun -->
  <joint name="edf1_left_mount" type="fixed">
    <child link="edf1_left"/>
    <parent link="gimbal1_pitch_module"/>
    <origin rpy="0 0 0" xyz="0 0.1115 0 "/>
    <axis xyz="0 0 1"/>
  </joint>
  <link name="edf1_left">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.00001"/>
      <inertia ixx="0.000001" ixy="0.0" ixz="0.0" iyy="0.000001" iyz="0.0" izz="0.000002"/>
    </inertial>
  </link>
  <joint name="edf1_right_mount" type="fixed">
    <child link="edf1_right"/>
    <parent link="gimbal1_pitch_module"/>
    <origin rpy="0 0 0" xyz="0 -0.1115 0 "/>
    <axis xyz="0 0 1"/>
  </joint>
  <link name="edf1_right">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.00001"/>
      <inertia ixx="0.000001" ixy="0.0" ixz="0.0" iyy="0.000001" iyz="0.0" izz="0.000002"/>
    </inertial>
  </link>
  <gazebo reference="main_body">
    <dampingFactor>0.00</dampingFactor>
  </gazebo>
  <gazebo reference="main_body">
    <dampingFactor>0.00</dampingFactor>
  </gazebo>
  <!-- hardware interface -->
  <transmission name="gimbal1_roll_tran">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="gimbal1_roll">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="gimbal_servo{id}_roll">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="gimbal1_pitch_tran">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="gimbal1_pitch">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="gimbal_servo{id}_pitch">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="rotor1_tran">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="rotor1">
      <hardwareInterface>RotorInterface</hardwareInterface>
    </joint>
    <actuator name="rotor1">
      <hardwareInterface>RotorInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="rotor3_tran">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="rotor3">
      <hardwareInterface>RotorInterface</hardwareInterface>
    </joint>
    <actuator name="rotor3">
      <hardwareInterface>RotorInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <!-- <xacro:friction self ="${self}"/> -->
  <!-- id:1 is spinal side, id:2 is another side-->
  <!-- gimbal roll -->
  <joint name="gimbal2_roll" type="revolute">
    <limit effort="3" lower="-3.14" upper="3.14" velocity="12.5"/>
    <parent link="main_body"/>
    <child link="gimbal2_roll_module"/>
    <origin rpy="0 0 0" xyz="-0.4638 0 0"/>
    <axis xyz="1 0 0"/>
    <dynamics damping="0.1" friction="0.0"/>
  </joint>
  <link name="gimbal2_roll_module">
    <inertial>
      <origin rpy="0 0 0" xyz="-0.0151 0.0012 0.0233"/>
      <mass value="0.3335"/>
      <inertia ixx="0.0008" ixy="0" ixz="-0.0002" iyy="0.0009" iyz="0" izz="0.0003"/>
    </inertial>
    <visual>
      <!-- <origin xyz="0 0 0" rpy="0 0 ${pi}"/> -->
      <origin rpy="0 0 3.1415926535897931" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://twin_hammer/urdf/mesh/gimbal_roll_module.dae" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>
    <!-- box collision not work good for FCL
     <collision>
       <origin rpy="0 0 0" xyz="${gimbal_roll_x_offset}  0 0"/>
       <geometry>
         <box size="0.103 0.064 0.156"/>
       </geometry>
     </collision>
     -->
    <!-- cylinder collision geometry works well for FCL -->
    <collision>
      <!-- <origin rpy="${pi/2} 0 0" xyz="0 0 0"/> -->
      <origin rpy="1.5707963267948965 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.3" radius="0.10"/>
      </geometry>
    </collision>
  </link>
  <!-- gimbal pitch -->
  <joint name="gimbal2_pitch" type="revolute">
    <limit effort="3" lower="-3.14" upper="3.14" velocity="12.5"/>
    <parent link="gimbal2_roll_module"/>
    <child link="gimbal2_pitch_module"/>
    <origin rpy="0 0 0" xyz="-0.0065 0 0.0265"/>
    <axis xyz="0 1 0"/>
    <dynamics damping="0.1" friction="0.0"/>
  </joint>
  <link name="gimbal2_pitch_module">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0.0005 -0.0092"/>
      <mass value="0.5276"/>
      <inertia ixx="0.0063" ixy="0.0" ixz="0.0" iyy="0.0004" iyz="0.0" izz="0.0062"/>
    </inertial>
    <visual>
      <!-- <origin xyz="0 0 0" rpy="0 0 ${pi}"/> -->
      <origin rpy="0 0 3.1415926535897931" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://twin_hammer/urdf/mesh/gimbal_pitch_module.dae" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>
  </link>
  <!-- rotor -->
  <joint name="rotor2" type="continuous">
    <limit effort="100.0" lower="2" upper="30" velocity="0.5"/>
    <parent link="gimbal2_pitch_module"/>
    <child link="thrust2"/>
    <origin rpy="0 0 0" xyz="0 0.1115 0"/>
    <axis xyz="0 0 1"/>
  </joint>
  <link name="thrust2">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.0001"/>
      <inertia ixx="0.00001" ixy="0.0" ixz="0.0" iyy="0.00001" iyz="0.0" izz="0.00002"/>
    </inertial>
  </link>
  <joint name="rotor4" type="continuous">
    <limit effort="100.0" lower="2" upper="30" velocity="0.5"/>
    <parent link="gimbal2_pitch_module"/>
    <child link="thrust4"/>
    <origin rpy="0 0 0" xyz="0 -0.1115 0"/>
    <axis xyz="0 0 1"/>
  </joint>
  <link name="thrust4">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.0001"/>
      <inertia ixx="0.00001" ixy="0.0" ixz="0.0" iyy="0.00001" iyz="0.0" izz="0.00002"/>
    </inertial>
  </link>
  <!-- special part of duct fun -->
  <joint name="edf2_left_mount" type="fixed">
    <child link="edf2_left"/>
    <parent link="gimbal2_pitch_module"/>
    <origin rpy="0 0 0" xyz="0 0.1115 0 "/>
    <axis xyz="0 0 1"/>
  </joint>
  <link name="edf2_left">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.00001"/>
      <inertia ixx="0.000001" ixy="0.0" ixz="0.0" iyy="0.000001" iyz="0.0" izz="0.000002"/>
    </inertial>
  </link>
  <joint name="edf2_right_mount" type="fixed">
    <child link="edf2_right"/>
    <parent link="gimbal2_pitch_module"/>
    <origin rpy="0 0 0" xyz="0 -0.1115 0 "/>
    <axis xyz="0 0 1"/>
  </joint>
  <link name="edf2_right">
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.00001"/>
      <inertia ixx="0.000001" ixy="0.0" ixz="0.0" iyy="0.000001" iyz="0.0" izz="0.000002"/>
    </inertial>
  </link>
  <gazebo reference="main_body">
    <dampingFactor>0.00</dampingFactor>
  </gazebo>
  <gazebo reference="main_body">
    <dampingFactor>0.00</dampingFactor>
  </gazebo>
  <!-- hardware interface -->
  <transmission name="gimbal2_roll_tran">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="gimbal2_roll">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="gimbal_servo{id}_roll">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="gimbal2_pitch_tran">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="gimbal2_pitch">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="gimbal_servo{id}_pitch">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="rotor2_tran">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="rotor2">
      <hardwareInterface>RotorInterface</hardwareInterface>
    </joint>
    <actuator name="rotor2">
      <hardwareInterface>RotorInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="rotor4_tran">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="rotor4">
      <hardwareInterface>RotorInterface</hardwareInterface>
    </joint>
    <actuator name="rotor4">
      <hardwareInterface>RotorInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <!-- <xacro:friction self ="${self}"/> -->
</robot>

